### ISSUE TITLE
[Bug]: react/hook-use-state incorrectly reporting overmind's useState

### ISSUE THREAD

**Issue by BrianHannay** [jsx-eslint/eslint-plugin-react#3540]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

I am trying out the `react/hook-use-state` rule on an existing project that uses a library that has a different useState hook which should not require destructuring.

Error:
```
/repo/pages/page.tsx
  26:17  error  useState call is not destructured into value + setter pair  react/hook-use-state
```

Code in question:
```
import React, { useEffect, ReactElement } from "react";
import { useActions, useState } from "../../../../overmind";
[...]
export default function Id(): ReactElement {
  const state = useState();
  const actions = useActions();
```

A useState function from [Overmind](https://overmindjs.org/v/v26/views/react) is being matched by the linting rule. Overmind's useState hook does not require destructuring.

`npx eslint .`

### Expected Behavior

I expected react/hook-use-state to only match react's `useState`. `useState` from other modules should not be evaluated by this linting rule.

I suspect this rule is only evaluated on function names matching `useState`. It may also be worth investigating whether this rule can be unintentionally bypassed by changing the name of the import. For example, this code should not pass linting:

```
import { useState as useAlteredState } from "react";
export default function Id(): ReactElement {
  const state = useAlteredState();
}
```

Expected no error for this code:
```
import React, { useEffect, ReactElement } from "react";
import { useActions, useState } from "../../../../overmind";
[...]
export default function Id(): ReactElement {
  const state = useState();
  const actions = useActions();
```

### eslint-plugin-react version

7.32.2

### eslint version

^8.34.0

### node version

v16.15.0

**Comment by ljharb** [comment]
While I agree this can *probably* be improved, and should be if so, a library choosing a hook name that collides with a primitive React hook seems like an exceedingly unwise decision that is likely to conflict with lots of things, including `eslint-plugin-react-hooks`.

**Comment by duncanbeevers** [comment]
The rule *should* support this differentiation.
I need a better reproduce case to make progress on this.

When testing with code similar to that reported in this issue, everything is passing.

```diff
diff --git a/tests/lib/rules/hook-use-state.js b/tests/lib/rules/hook-use-state.js
index 1cadaf8a..398fb94e 100644
--- a/tests/lib/rules/hook-use-state.js
+++ b/tests/lib/rules/hook-use-state.js
@@ -105,6 +105,24 @@ const tests = {
         const [color, setFlavor] = useState()
       `,
     },
+    // These tests pass with no code changes
+    {
+      code: `
+        import React from 'react'
+        import { useActions, useState } from 'overmind'
+        const overmindState = useState()
+      `,
+    },
+    {
+      code: `
+        import React, { useEffect, ReactElement } from 'react'
+        import { useActions, useState } from 'overmind'
+        export default function Id(): ReactElement {
+          const actions = useActions()
+          const state = useState()
+        }
+      `,
+      features: ['ts'],
+    },
     {
       code: `
         import { useRef } from 'react'
```

**Comment by ljharb** [comment]
@duncanbeevers presumably it’d only detect it inside a react component, which must be necessity be returning jsx (or something renderable). Your second test case where it returns jsx might work?

**Comment by duncanbeevers** [comment]
The checks I put in place don't try and detect the context in which the `useState` call is made.
They only check that the call is to the function exported as `useState` from the `react` module.

Imports from React are tracked in `lib/util/Components.js`, and hook calls are identified using the [`isReactHookCall`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/util/Components.js#L753-L823) check.

The context in which those imports are used is *not* considered by this rule.
The issue here appears to be that this non-React `useState` call is being identified as a React hook call, which implies some bug within this `isReactHookCall` method.

However, both test cases I provide pass, correctly identifying and ignorning both of these non-React `useState` calls.
I cannot get this rule to fail in the way the OP describes.

**Comment by duncanbeevers** [comment]
> presumably it’d only detect it inside a react component, which must be necessity be returning jsx (or something renderable). Your second test case where it returns jsx might work?

As we discussed this before, the issue with attempting to detect these calls only within React components is that such a strategy precludes detecting these calls within custom hooks.

```ts
function useMyHook() {
  const [myState, setMyState] = useState('wow');
  return { state: myState, onSetState: setMyState };
}
```

**Comment by ljharb** [comment]
Thanks for the reminder :-)

@brianhannay is this still an issue for you? At the moment it seems like we can’t reproduce it.

**Comment by duncanbeevers** [comment]
@ljharb I'd like to close this issue out, since there's no reproduction.
